webhook/resources/volume/validator.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/longhorn/longhorn-manager/util"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+)
+
+type volumeValidator struct {
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &volumeValidator{ds: ds}
+}
+
+func (v *volumeValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "volumes",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.Volume{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (v *volumeValidator) Create(request *admission.Request, newObj runtime.Object) error {
+	volume := newObj.(*longhorn.Volume)
+
+	if !util.ValidateName(volume.Name) {
+		return werror.NewInvalidError(fmt.Sprintf("invalid name %v", volume.Name), "")
+	}
+
+	if err := types.ValidateDataLocality(volume.Spec.DataLocality); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := types.ValidateAccessMode(volume.Spec.AccessMode); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := types.ValidateReplicaCount(volume.Spec.NumberOfReplicas); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := types.ValidateReplicaAutoBalance(volume.Spec.ReplicaAutoBalance); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if volume.Spec.BackingImage != "" {
+		if _, err := v.ds.GetBackingImage(volume.Spec.BackingImage); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if volume.Spec.EngineImage == "" {
+		return werror.NewInvalidError("BUG: Invalid empty Setting.EngineImage", "")
+	}
+
+	if !volume.Spec.Standby {
+		if volume.Spec.Frontend != longhorn.VolumeFrontendBlockDev && volume.Spec.Frontend != longhorn.VolumeFrontendISCSI {
+			return werror.NewInvalidError(fmt.Sprintf("invalid volume frontend specified: %v", volume.Spec.Frontend), "")
+		}
+	}
+
+	if volume.Spec.Migratable && volume.Spec.AccessMode != longhorn.AccessModeReadWriteMany {
+		return werror.NewInvalidError("migratable volumes are only supported in ReadWriteMany (rwx) access mode", "")
+	}
+
+	// Check engine version before disable revision counter
+	if volume.Spec.RevisionCounterDisabled {
+		if ok, err := v.canDisableRevisionCounter(volume.Spec.EngineImage); !ok {
+			err := errors.Wrapf(err, "can not create volume with current engine image that doesn't support disable revision counter")
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	currentNodeID, err := util.GetRequiredEnv(types.EnvNodeName)
+	if err != nil {
+		return werror.NewInvalidError("BUG: fail to detect the node name", "")
+	}
+
+	m := manager.NewVolumeManager(currentNodeID, v.ds)
```

Comment:
IMO, pkg `manager` is mainly used by the functions in pkg `api`. You can move the required util functions to other pkgs like `datastore` instead.